applications/HDF5Application/custom_utilities/factor_elements_and_conditions_utility.cpp

```diff
@@ -0,0 +1,274 @@
+std::vector<std::string> GetLocalComponentNames(const TContainerType& rElements)
+{
+    KRATOS_TRY;
+
+    if (rElements.size() == 0)
+        return std::vector<std::string>();
+
+    std::vector<std::pair<std::string, const typename TContainerType::data_type *>> components(1);
+    std::string name;
+    // Add first component.
+    CompareElementsAndConditionsUtility::GetRegisteredName(rElements.front(), name);
+    components[0] = std::pair<std::string, const typename TContainerType::data_type *>(
+        name, &rElements.front());
+
+    // Find all remaining components.
+    components.reserve(5);
+    unsigned pos = 0;
+    for (typename TContainerType::const_iterator it = rElements.begin() + 1;
+         it != rElements.end(); ++it)
+    {
+        if (GeometricalObject::IsSame(*it, *components[pos].second) == false)
+        {
+            bool found = false;
+            for (unsigned k = 0; k < components.size(); ++k)
+            {
+                if (GeometricalObject::IsSame(*it, *components[k].second))
+                {
+                    found = true;
+                    pos = k;
+                    break;
+                }
+            }
+
+            if (!found)
+            {
+                CompareElementsAndConditionsUtility::GetRegisteredName(*it, name);
+                components.push_back(std::pair<std::string, const typename TContainerType::data_type*>(
+                    name, &(*it)));
+                pos = components.size() - 1;
+            }
+        }
+    }
+
+    // Return component names.
+    std::vector<std::string> component_names(components.size());
+    for (unsigned i = 0; i < components.size(); ++i)
+        component_names[i] = components[i].first;
+    return component_names;
+
+    KRATOS_CATCH("");
+}
+
+template<class TContainerType>
+// ElementContainerType || ConditionContainerType
+std::vector<std::string> GetComponentNames(const TContainerType& rElements)
+{
+    KRATOS_TRY;
+
+    const int max_size = 1000;
+    std::vector<std::string> local_component_names =
+        GetLocalComponentNames<TContainerType>(rElements);
+
+    // Construct global vector of component names across all processes if MPI is used.
+    std::vector<std::string> component_names;
+#ifdef KRATOS_USING_MPI
+    int mpi_is_initialized, ierr;
+    ierr = MPI_Initialized(&mpi_is_initialized);
+    KRATOS_ERROR_IF(ierr != MPI_SUCCESS) << "MPI_Initialized failed." << std::endl;
+
+    if (mpi_is_initialized)
+    {
+        const int root = 0;
+        int comm_rank, comm_size, ierr;
+        MPI_Comm_rank(MPI_COMM_WORLD, &comm_rank);
+        MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
+
+        // Fill the send buffer with local component names.
+        std::stringstream ss;
+        for (const auto& r_name : local_component_names)
+            ss << r_name << '\n';
+        const std::string send_buf = ss.str();
+
+        // Gather components from each process and construct the global set of
```

Comment:
If I understand the next loop correctly you are doing some sort of [allreduce](https://www.mpich.org/static/docs/v3.1/www3/MPI_Reduce.html)
First you gather and then you scatter again
Couldn't you use the mpi-cmd directly? It might not be allreduce but I think there should be sth for this.

Reply:
I think it is closest to a gather, but with gather we need to have all the messages the same size.